tests/devices/test_default_qubit_jax.py

```diff
@@ -0,0 +1,462 @@
+        dev = qml.device("default.qubit.jax", wires=2)
+        assert dev.num_wires == 2
+        assert dev.shots == 1000
+        assert dev.analytic
+        assert dev.short_name == "default.qubit.jax"
+        assert dev.capabilities()["passthru_interface"] == "jax"
+
+    def test_qubit_circuit(self, tol):
+        """Test that the device provides the correct
+        result for a simple circuit."""
+        p = jnp.array(0.543)
+
+        dev = qml.device("default.qubit.jax", wires=1)
+
+        @qml.qnode(dev, interface="jax")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliY(0))
+
+        expected = -jnp.sin(p)
+        if not qml.tape_mode_active():
+            assert isinstance(circuit, qml.qnodes.PassthruQNode)
+        assert jnp.isclose(circuit(p), expected, atol=tol, rtol=0)
+
+    def test_qubit_circuit_with_jit(self, tol):
+        """Test that the device provides the correct
+        result for a simple circuit."""
+        p = jnp.array(0.543)
+
+        dev = qml.device("default.qubit.jax", wires=1)
+
+        @jax.jit
+        @qml.qnode(dev, interface="jax")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliY(0))
+
+        expected = -jnp.sin(p)
+        # Do not test isinstance here since the @jax.jit changes the function
+        # type.
+        # Just test that it works and spits our the right value.
+        assert jnp.isclose(circuit(p), expected, atol=tol, rtol=0)
+
+    def test_correct_state(self, tol):
+        """Test that the device state is correct after applying a
+        quantum function on the device"""
+
+        dev = qml.device("default.qubit.jax", wires=2)
+
+        state = dev.state
+        expected = jnp.array([1, 0, 0, 0])
+        assert jnp.allclose(state, expected, atol=tol, rtol=0)
+
+        @qml.qnode(dev, interface="jax", diff_method="backprop")
+        def circuit():
+            qml.Hadamard(wires=0)
+            qml.RZ(jnp.pi / 4, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        circuit()
+        state = dev.state
+
+        amplitude = jnp.exp(-1j * jnp.pi / 8) / jnp.sqrt(2)
+
+        expected = jnp.array([amplitude, 0, jnp.conj(amplitude), 0])
+        assert jnp.allclose(state, expected, atol=tol, rtol=0)
+
+    def test_sampling_with_jit(self):
+        """Test that sampling works with a jax.jit"""
+        @jax.jit
+        def circuit(key):
+            dev = qml.device("default.qubit.jax", wires=1, prng_key=key)
+            @qml.qnode(dev, interface="jax", diff_method="backprop")
+            def inner_circuit():
+                qml.Hadamard(0)
+                return qml.sample(qml.PauliZ(wires=0))
+            return inner_circuit()
+
+        a = circuit(jax.random.PRNGKey(0))
+        b = circuit(jax.random.PRNGKey(0))
+        c = circuit(jax.random.PRNGKey(1))
+        np.testing.assert_array_equal(a, b)

```

Comment:
Interesting, didn't know about `np.testing` before!

Reply:
How come you went with that in this case?

Reply:
Force of habit. `assert (a == b).all()` would also work, but the numpy testing module gives you nice extra information when a test fails.

Example:
```python
>>> import numpy as np
>>> a = np.array([1, 1, 0])
>>> b = np.array([1, 1, 1])
>>> np.testing.assert_array_equal(a, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/chase/anaconda3/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 930, in assert_array_equal
    assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,
  File "/home/chase/anaconda3/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 840, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Arrays are not equal

Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 1
Max relative difference: 1.
 x: array([1, 1, 0])
 y: array([1, 1, 1])
```

Reply:
Oh I didn't know about the NumPy testing module either! Looks awesome.

Reply:
Yeah its really nice. It will also work for torch and tensorflow type tensors! We used it a lot in TN.